### PR TITLE
Implemented Getting Zenodo citation

### PR DESCRIPTION
**Pull request by davibarreira**
I've implemented two new functions in `citations.jl`:
`get_zenodo_badge` - This function reads the `README.md` file and extracts the link from the zenodo's badge;
`get_citation_zenodo` - This functions takes the badge link and obtains the bibtex.

I've modified the `get_citations` function to search for the badge in case no `Citation.bib` file is found. I've also added an argument `zenodo=false`  to `get_citation`, `get_tool_citations` and `collect_citations`. The default is `false` so that your previous tests do not break.

I've also implemented a testset for the zenodo functions. Note that some of your tests were already failing, and they are not related to the new functions.

I hope this PR is helpful.

### CODE REVIEW

File: src/citations.jl [hunk 1]
```diff
@@ -5,32 +5,43 @@ function citation_path(pkg)
     end
 end
 
-function get_citation(pkg)
+
+# Added `zenodo` flag to avoid breaking current tests
+function get_citation(pkg; zenodo=false)
```
Comment (SebastianM-C) [inline src/citations.jl:10]: I think that it might be better to have a more general flag like `badge` (or `include_badges` or something like that).

**Review by SebastianM-C — CHANGES_REQUESTED** [review]
Thank you for your contribution!

I left some comments and suggestions on the specific parts of the code. It looks good, but I think it would be better to refactor the code a bit to return a merged citation dictionary from `get_citation` and maybe have a function `get_badge_citations` that would call `get_citation_zenodo`. This way it will be easier to extend for JOSS (and other types) of badges. I also think that we should default to getting all the citation data.

I'm wondering if we could just setup [a regex for DOIs](https://stackoverflow.com/questions/27910/finding-a-doi-in-a-document-or-page) in badges.

> Note that some of your tests were already failing, and they are not related to the new functions.

Can you provide some information about the failing tests? The CI doesn't show failing package tests and I also checked locally and I didn't encounter any problems.
The errors in CI like
```
Error: Can't open display: (null)
```
are due to the clipboard integration tests because I haven't managed to install `xclip` properly (#11).

File: src/citations.jl [hunk 2]
```diff
@@ -5,32 +5,43 @@ function citation_path(pkg)
     end
 end
 
-function get_citation(pkg)
+
+# Added `zenodo` flag to avoid breaking current tests
```
[the remark this answers is not in the document]
Comment (davibarreira) [reply src/citations.jl:9]: I agree, but I thought this should be done in another PR, cause it would be too much change.

**Comment by davibarreira** [comment]
> Thank you for your contribution!
> 
> I left some comments and suggestions on the specific parts of the code. It looks good, but I think it would be better to refactor the code a bit to return a merged citation dictionary from `get_citation` and maybe have a function `get_badge_citations` that would call `get_citation_zenodo`. This way it will be easier to extend for JOSS (and other types) of badges. I also think that we should default to getting all the citation data.
> 
> I'm wondering if we could just setup [a regex for DOIs](https://stackoverflow.com/questions/27910/finding-a-doi-in-a-document-or-page) in badges.
> 
> > Note that some of your tests were already failing, and they are not related to the new functions.
> 
> Can you provide some information about the failing tests? The CI doesn't show failing package tests and I also checked locally and I didn't encounter any problems.
> The errors in CI like
> 
> ```
> Error: Can't open display: (null)
> ```
> 
> are due to the clipboard integration tests because I haven't managed to install `xclip` properly (#11).

What exactly do you mean by using "regex for DOI"? Do you mean, search for the "DOI" term, and then extract the badge link? I guess this could be done, although I'm not very good with regex.

File: src/citations.jl [hunk 3]
```diff
@@ -5,32 +5,43 @@ function citation_path(pkg)
...
+    end
     if !isnothing(bib_path)
         @debug "Reading CITATION.bib for $(pkg.name)"
         try
             import_bibtex(bib_path)
         catch e
             @warn "There was an error reading the CITATION.bib file for $(pkg.name)" exception=e
         end
+    elseif !isnothing(urlzenodo)
+        get_citation_zenodo(urlzenodo)
```
[the remark this answers is not in the document]
Comment (davibarreira) [reply src/citations.jl:25]: The way it's now, it just get's the zenodo citation if no CITATION.bib is not present. Isn't this your suggestion?

File: src/PkgCite.jl [hunk 4]
```diff
@@ -2,13 +2,15 @@ module PkgCite
 
 using Pkg: include
 using DataStructures: getkey, values
-using Base: String
+using Base: String, indent_width
```
[the remark this answers is not in the document]
Comment (davibarreira) [reply src/PkgCite.jl:5]: I forgot to remove it. It's not used.

File: src/citations.jl [hunk 5]
```diff
@@ -80,3 +93,49 @@ function get_citations(;only_direct=false, filename="julia_citations.bib")
...
+    get_citation_zenodo(url::String)
+
+This function will return an `OrderedDict` in the default to BibTeX format
+from `Bibliography.jl`. The `url` argument is the link present in the
+Zenodo badge.
+"""
+function get_citation_zenodo(url)
+    if url == nothing
+        return nothing
+    end
```
[the remark this answers is not in the document]
Comment (davibarreira) [reply src/citations.jl:107]: Didn't know, thanks.

File: src/citations.jl [hunk 6]
```diff
@@ -80,3 +93,49 @@ function get_citations(;only_direct=false, filename="julia_citations.bib")
...
+"""
+    get_zenodo_badge(pkg)
+
+This function extracts the link in the Zenodo badge
+present in the package's `README.md`.
+"""
+function get_zenodo_badge(pkg)
+    readme_path = joinpath(pkg.source, "README.md")
+    if isfile(readme_path)
+        readme = open(f->read(f, String), readme_path);
```
[the remark this answers is not in the document]
Comment (davibarreira) [reply src/citations.jl]: Leftover from jupyter-notebook. I use the `;` to hide the output.

**Comment by davibarreira** [comment]
By the way, here are the test failures I get when I just run the forked repo without my changes:
```julia
 [0c5d862f] Symbolics v0.1.27
┌ Warning: There was an error reading the CITATION.bib file for Distributions
│   exception = KeyError: key "july" not found
└ @ PkgCite ~/MEGA/EMAp/PkgCite.jl/src/citations.jl:15
┌ Warning: There was an error reading the CITATION.bib file for Distributions
│   exception = KeyError: key "july" not found
└ @ PkgCite ~/MEGA/EMAp/PkgCite.jl/src/citations.jl:15
┌ Warning: Overwriting julia_citations.bib
└ @ PkgCite ~/MEGA/EMAp/PkgCite.jl/src/citations.jl:53
[ Info: A julia_citations.bib file with the citations for LabelledArrays, QuadGK, Symbolics, RecursiveArrayTools, ArrayInterface and AbstractAlgebra was generated in the current working directory (/home/davibarreira/MEGA/EMAp/PkgCite.jl/test).
┌ Warning: There was an error reading the CITATION.bib file for Distributions
│   exception = KeyError: key "july" not found
└ @ PkgCite ~/MEGA/EMAp/PkgCite.jl/src/citations.jl:15
┌ Warning: There was an error reading the CITATION.bib file for Distributions
│   exception = KeyError: key "july" not found
└ @ PkgCite ~/MEGA/EMAp/PkgCite.jl/src/citations.jl:15
[ Info: The above sentence was copied to your clipboard.
┌ Warning: Overwriting julia_citations.bib
└ @ PkgCite ~/MEGA/EMAp/PkgCite.jl/src/citations.jl:53
[ Info: A julia_citations.bib file with the citations for LabelledArrays, QuadGK, Symbolics, RecursiveArrayTools, Julia, ArrayInterface and AbstractAlgebra was generated in the current working directory (/home/davibarreira/MEGA/EMAp/PkgCite.jl/test).
Clipboard: Error During Test at /home/davibarreira/MEGA/EMAp/PkgCite.jl/test/runtests.jl:81
 Unexpected Pass
 Expression: clipboard() == CITE_STR_JL
 Got correct result, please change to @test if no longer broken.

┌ Warning: There was an error reading the CITATION.bib file for Distributions
│   exception = KeyError: key "july" not found
└ @ PkgCite ~/MEGA/EMAp/PkgCite.jl/src/citations.jl:15
[ Info: The above sentence was copied to your clipboard.
┌ Warning: Overwriting julia_citations.bib
└ @ PkgCite ~/MEGA/EMAp/PkgCite.jl/src/citations.jl:53
[ Info: A julia_citations.bib file with the citations for LabelledArrays, QuadGK, Symbolics, RecursiveArrayTools, Julia, ArrayInterface and AbstractAlgebra was generated in the current working directory (/home/davibarreira/MEGA/EMAp/PkgCite.jl/test).
Clipboard: Error During Test at /home/davibarreira/MEGA/EMAp/PkgCite.jl/test/runtests.jl:89
 Unexpected Pass
 Expression: clipboard() == CITE_STR
 Got correct result, please change to @test if no longer broken.

Test Summary:              | Pass  Error  Broken  Total
PkgCite.jl                 |   24      2       1     27
  Empty env                |    1                     1
  Packages                 |    6              1      7
  Citations                |    4                     4
  Citations in .bib        |    5                     5
  Only direct dependencies |    3                     3
  PkgCite sentence         |    5      2              7
    Clipboard              |    2      2              4
ERROR: LoadError: Some tests did not pass: 24 passed, 0 failed, 2 errored, 1 broken.
in expression starting at /home/davibarreira/MEGA/EMAp/PkgCite.jl/test/runtests.jl:10
ERROR: Package PkgCite errored during testing
```

**Review by SebastianM-C — COMMENTED** [review]
> Also, can you give an example of what do you mean by the " but I think it would be better to refactor the code a bit to return a merged citation dictionary from get_citation and maybe have a function get_badge_citations that would call get_citation_zenodo"? I don't fully understand what exactly would be this merged citation dictionary.

I was thinking if there would be an edge case where we need to merge citations form the badge with the ones in the CITATION.bib file, but as I've said it the response comment, it would be better to just document that we prefer the CITATION.bib file over the badges and leave the implementation as is.

> What exactly do you mean by using "regex for DOI"? Do you mean, search for the "DOI" term, and then extract the badge link? I guess this could be done, although I'm not very good with regex.

I was referring to parsing the README for badges and then look for DOI-like strings (things like "10.5281/ZENODO.4692172") insde them since the badge links contain the DOI information, but I see now that this will not work for Zenodo since there can be links with `latestdoi` that need to be followed for the DOI. I thought that the DOI info is present in the link like in [the badge for Plots](https://raw.githubusercontent.com/JuliaPlots/Plots.jl/master/README.md) or [DynamicalSystems](https://raw.githubusercontent.com/JuliaDynamics/DynamicalSystems.jl/master/README.md).

Looking for the "DOI" string is indeed a better solution since it seems to be used in all the badges I've encountered.

**Comment by davibarreira** [comment]
> > Also, can you give an example of what do you mean by the " but I think it would be better to refactor the code a bit to return a merged citation dictionary from get_citation and maybe have a function get_badge_citations that would call get_citation_zenodo"? I don't fully understand what exactly would be this merged citation dictionary.
> 
> I was thinking if there would be an edge case where we need to merge citations form the badge with the ones in the CITATION.bib file, but as I've said it the response comment, it would be better to just document that we prefer the CITATION.bib file over the badges and leave the implementation as is.
> 
> > What exactly do you mean by using "regex for DOI"? Do you mean, search for the "DOI" term, and then extract the badge link? I guess this could be done, although I'm not very good with regex.
> 
> I was referring to parsing the README for badges and then look for DOI-like strings (things like "10.5281/ZENODO.4692172") insde them since the badge links contain the DOI information, but I see now that this will not work for Zenodo since there can be links with `latestdoi` that need to be followed for the DOI. I thought that the DOI info is present in the link like in [the badge for Plots](https://raw.githubusercontent.com/JuliaPlots/Plots.jl/master/README.md) or [DynamicalSystems](https://raw.githubusercontent.com/JuliaDynamics/DynamicalSystems.jl/master/README.md).
> 
> Looking for the "DOI" string is indeed a better solution since it seems to be used in all the badges I've encountered.

Yeah, there are some cases where the DOI is in the badge, but for Zenodo, at least, this is not the correct way to setup. The reason is that for every new package version a new DOI is generated. So using "latestdoi" is the correct setup, hence, why I wrote that "warning". Perhaps, one can check if the badge has a DOI, and use that instead.

**Comment by SebastianM-C** [comment]
Regarding the failing tests, I marked the `clipboard` ones as broken since that functionality is broken on CI (and it didn't work on the local computer where I tested because `xclip`/`xsel` wasn't installed), but it worked on your computer. I think I will wrap those tests in a `try` that checks `clipboard()` works in a future PR. We can safely ignore this for now.

**Comment by SebastianM-C** [comment]
> Yeah, there are some cases where the DOI is in the badge, but for Zenodo, at least, this is not the correct way to setup. The reason is that for every new package version a new DOI is generated. So using "latestdoi" is the correct setup, hence, why I wrote that "warning". Perhaps, one can check if the badge has a DOI, and use that instead.

It looks like looking for a badge that starts with "DOI" will cover all cases (and also work for JOSS badges). We could issue a warning in the case of Zenodo if we detect a badge that doesn't use `latestdoi`, but using a DOI for all the package versions (like Plots) or the latest one seems like something that not everyone will agree on, so we should test that the code works on both variants.

**Comment by davibarreira** [comment]
I'll do the following. I'll create a function that searchers for a badge with DOI, and get's the link. Now, if the link if from zenodo, then uses my code to get the bibtex. If it's a doi link, then I'll create another function to generate the bibtex from it.

**Comment by davibarreira** [comment]
Made all the suggested changes. The function `get_zenodo_badge` now is `get_badge`, and it looks for the DOI badge, and extracts the link. The function `get_citation_zenodo` is now `get_citation_badge` and it checks weather the link is already the DOI or if it's the `latestdoi` (for the Zenodo case). I guess this is all.
As I said in the other comment, I left the default as `false` and I think a new PR should be opened to turn `true` into the default, improve the documentation and fix the tests.

**Review by SebastianM-C — APPROVED** [review]
Thank you very much for your contribution!

File: test/cite_str.jl [hunk 7]
```diff
@@ -34,6 +34,7 @@ const CITE_STR =
 "ArrayInterface\\cite[ArrayInterface]{DifferentialEquations.jl-2017} and "*
 "AbstractAlgebra\\cite[AbstractAlgebra]{AbstractAlgebra.jl-2017}."
 
-const CITE_ZENODO =
+const CITE_BADGE =
 "This work was done in \\cite[Julia v$VERSION]{Julia-2017} "*
-"and made use of the WriteVTK.jl\\cite[WriteVTK]{https://doi.org/10.5281/zenodo.4692172} package."
\ No newline at end of file
+"and made use of the Plots.jl\\cite[Plots]{https://doi.org/10.5281/zenodo.4725317} "*
+"and WriteVTK.jl\\cite[WriteVTK]{https://doi.org/10.5281/zenodo.4692172} packages."
```
Comment (SebastianM-C) [inline test/cite_str.jl:40]:
```suggestion
"and WriteVTK.jl\\cite[WriteVTK]{https://doi.org/10.5281/zenodo.4692172} packages."

```

**Comment by davibarreira** [comment]
> Thank you very much for your contribution!

Not a problem! Thanks for the pkg by the way :D

**Pull request merged**